backend/repository/validator.go

```diff
@@ -0,0 +1,378 @@
+		err = errors.New("empty queue")
+		return
+	}
+	node := nq.q[0]
+	client = node.ClientNode
+	endpoint = node.EndpointNode
+	nq.q = nq.q[1:]
+	return
+}
+
+// Empty returns true if queue is empty
+func (nq NodeQ) Empty() bool {
+	return len(nq.q) == 0
+}
+
+// SyntaxVisitor implements a syntax compare visitor interface
+type SyntaxVisitor struct {
+	nodeQ *NodeQ
+}
+
+// NewSyntaxVisitor returns a SyntaxVisitor
+func NewSyntaxVisitor(clientNode, endpointNode interface{}) *SyntaxVisitor {
+	nodeQ := NewNodeQ()
+	nodeQ.EnQ(clientNode, endpointNode)
+	return &SyntaxVisitor{
+		nodeQ: nodeQ,
+	}
+}
+
+// Visit BFS visits all node pairs in node queue
+func (sv *SyntaxVisitor) Visit() error {
+	var (
+		client   interface{}
+		endpoint interface{}
+		err      error
+	)
+	for {
+		client, endpoint, err = sv.nodeQ.DeQ()
+		if err != nil {
+			return nil
+		}
+		if reflect.TypeOf(client) != reflect.TypeOf(endpoint) {
+			return errors.New("type mismatch")
+		}
+		if err := visit(sv, client, endpoint); err != nil {
+			return err
+		}
+	}
+}
+
+func visit(vs Visitor, client, endpoint interface{}) error {
+	switch endpoint.(type) {
+	case *compile.ServiceSpec:
+		return vs.VisitService(client.(*compile.ServiceSpec), endpoint.(*compile.ServiceSpec))
+	case *compile.FunctionSpec:
+		return vs.VisitFunction(client.(*compile.FunctionSpec), endpoint.(*compile.FunctionSpec))
+	case compile.ArgsSpec:
+		return vs.VisitArgsSpec(client.(compile.ArgsSpec), endpoint.(compile.ArgsSpec))
+	case *compile.ResultSpec:
+		return vs.VisitResultSpec(client.(*compile.ResultSpec), endpoint.(*compile.ResultSpec))
+	case compile.FieldGroup:
+		return vs.VisitFieldGroup(client.(compile.FieldGroup), endpoint.(compile.FieldGroup))
+	case *compile.FieldSpec:
+		return vs.VisitFieldSpec(client.(*compile.FieldSpec), endpoint.(*compile.FieldSpec))
+	case compile.TypeSpec:
+		return vs.VisitTypeSpec(client.(compile.TypeSpec), endpoint.(compile.TypeSpec))
+	default:
+		return errors.New("unsupported type")
+	}
+}
+
+// VisitService visit and syntax check client/endpoint service
+func (sv *SyntaxVisitor) VisitService(*compile.ServiceSpec, *compile.ServiceSpec) error {
+	return nil
+}
+
+// VisitFunction visit and syntax check client/endpoint function
+func (sv *SyntaxVisitor) VisitFunction(client *compile.FunctionSpec, endpoint *compile.FunctionSpec) error {
+	if client.Name != endpoint.Name {
+		return NewRequestError(
+			ThriftFiles,
+			fmt.Errorf("function name mismatch: %s in endpoint but %s in client", endpoint.Name, client.Name))
+	}
+	sv.nodeQ.EnQ(client.ArgsSpec, endpoint.ArgsSpec)
+	sv.nodeQ.EnQ(client.ResultSpec, endpoint.ResultSpec)
+	return nil
+}
+
+// VisitArgsSpec visit and syntax check client/endpoint argsSpec
+func (sv *SyntaxVisitor) VisitArgsSpec(clientArgs compile.ArgsSpec, endpointArgs compile.ArgsSpec) error {
+	clientArgList := compile.FieldGroup(clientArgs)
+	endpointArgList := compile.FieldGroup(endpointArgs)
+	if len(clientArgList) != len(endpointArgList) {
```

Comment:
this check is too strict, the client can have extra (optional) arguments.

In fact we convert arguments by argument name so its possible for them to have the same length but different names.